### PR TITLE
Write CI in jsonnet

### PR DESCRIPTION
**Pull request by jolheiser**
This is a draft implementation of using [jsonnet](https://jsonnet.org/) to write the CI pipelines, and then convert them to YAML for Drone to consume. 
Technically Drone can consume jsonnet as well, however it seemed easier and less risky to end up with a static file for Drone to consume rather than have to evaluate jsonnet each time.

Some stats:
Current [drone file](https://github.com/go-gitea/gitea/blob/29b78bce3336fcedc08d17b71a52e593d65cd951/.drone.yml#L1509): 1,509 lines
This PR jsonnet file:  1,440 lines
This PR drone file: 1,374 lines (because it removes spacing)

This should make updating the CI a little less painful as we can abstract some of the duplication into functions as well as only having to update e.g. the Go version in a single spot to update CI without worrying about missing a spot.

This is the first step towards updating CI as a baseline.
This PR, in itself, **should not** change the current pipelines.

-----

> Okay, but clearly `.drone.yml` is changed

The contrib in this PR serves two functions:
1. Actually evaluating the jsonnet and outputting static YAML
2. A `--compare` mode in the event someone makes a change and wants to ensure the resulting output hasn't changed

I used the latter prior to generating the `.drone.yml` in this PR diff (and you should as well to check!) to make sure the output document is the same.

### CODE REVIEW

**Comment by silverwind** [comment]
> This should make updating the CI a little less painful

I'm not sure but I never heard/used jsonnet, so it'd be more "painful" having to learn another "language" 😉

**Comment by silverwind** [comment]
Personally, I prefer YAML over any JSON variants because it's less verbose where quotes are in most cases unneccesary.

**Comment by jolheiser** [comment]
I am open to alternatives or suggestions.
This PR mostly aims at trying to control our CI pipelines a bit more and trying to handle the "almost the same with a few tweaks" parts.

**Comment by silverwind** [comment]
>  This PR mostly aims at trying to control our CI pipelines a bit more and trying to handle the "almost the same with a few tweaks" parts.

Tangentially related: Having worked with JSON schema a lot, I came to like [refs](http://json-schema.org/understanding-json-schema/structuring.html#ref) as a way to de-duplicate documents, but even if YAML has something like it, it is not possible in drone files because they are essentially multiple documents.

**Comment by jolheiser** [comment]
Yeah, YAML has anchors and a few other things, however unfortunately some of our tweaks are a bit more than refs or anchors provide for. 
It is up to the individual to determine if the tradeoff is worth it, but for example [dockerLinuxRelease](https://github.com/jolheiser/gitea/blob/dc6524244444d7d7d18ee2d932e4f19a3a588429/.drone.jsonnet#L49-L139) is used [six times](https://github.com/jolheiser/gitea/blob/dc6524244444d7d7d18ee2d932e4f19a3a588429/.drone.jsonnet#L1435-L1441) with various parameters.  
Those parameters do change the output, and for someone unused to jsonnet it might take a second to click, but imo being able to modify one function to update six pipelines is a decent tradeoff.

-----

Just to put them out here, the other langs I looked at prior to jsonnet were:
[starlark](https://github.com/bazelbuild/starlark)
[cue](https://github.com/cue-lang/cue)

The reason I stuck with jsonnet is partially because I figured JSON was likely to be something people were more familiar with, and also because I would like to use jsonnet for custom/generic webhooks in the future, so the dep would be re-usable.

**Comment by techknowlogick** [comment]
> Yeah, YAML has anchors and a few other things, however unfortunately some of our tweaks are a bit more than refs or anchors provide for.

To go into more details about this, is that the parsing that drone employs means that yaml anchors aren't able to cross yaml document lines. In the case of drone, it means each pipeline, even though they are in the same file.

File: .drone.jsonnet [hunk 1]
```diff
@@ -0,0 +1,1445 @@
+// Images
+local minGoImage = 'golang:1.18';
+local goImage = 'golang:1.19';
+local xgoImage = 'techknowlogick/xgo:go-1.19.x';
```
Comment (techknowlogick) [inline .drone.jsonnet]:
```suggestion
local minGoImage = 'golang:1.19';
local goImage = 'golang:1.20';
local xgoImage = 'techknowlogick/xgo:go-1.20.x';
```

File: .drone.jsonnet [hunk 2]
```diff
@@ -0,0 +1,1445 @@
...
+        },
+      },
+      image: 'woodpeckerci/plugin-s3:latest',
+      name: 'release-branch',
+      pull: 'always',
+      settings: {
+        acl: 'public-read',
+        bucket: 'gitea-artifacts',
+        endpoint: 'https://ams3.digitaloceanspaces.com',
+        path_style: true,
```
Comment (techknowlogick) [inline .drone.jsonnet]:
```suggestion
        acl: {
                from_secret: 'aws_s3_acl',
        },
        region: {
                from_secret: 'aws_s3_region',
        },
        bucket: {
                from_secret: 'aws_s3_bucket',
        },
        endpoint: {
                from_secret: 'aws_s3_endpoint',
        },
        path_style: {
                from_secret: 'aws_s3_path_style',
        },
```

File: .drone.jsonnet [hunk 3]
```diff
@@ -0,0 +1,1445 @@
...
+          from_secret: 'aws_secret_access_key',
+        },
+      },
+      image: 'woodpeckerci/plugin-s3:latest',
+      name: 'release-main',
+      settings: {
+        acl: 'public-read',
+        bucket: 'gitea-artifacts',
+        endpoint: 'https://ams3.digitaloceanspaces.com',
+        path_style: true,
```
Comment (techknowlogick) [inline .drone.jsonnet]:
```suggestion
        acl: {
                from_secret: 'aws_s3_acl',
        },
        region: {
                from_secret: 'aws_s3_region',
        },
        bucket: {
                from_secret: 'aws_s3_bucket',
        },
        endpoint: {
                from_secret: 'aws_s3_endpoint',
        },
        path_style: {
                from_secret: 'aws_s3_path_style',
        },
```

File: .drone.jsonnet [hunk 4]
```diff
@@ -0,0 +1,1445 @@
...
+        },
+      },
+      image: 'woodpeckerci/plugin-s3:latest',
+      name: 'release-tag',
+      pull: 'always',
+      settings: {
+        acl: 'public-read',
+        bucket: 'gitea-artifacts',
+        endpoint: 'https://ams3.digitaloceanspaces.com',
+        path_style: true,
```
Comment (techknowlogick) [inline .drone.jsonnet]:
```suggestion
        acl: {
                from_secret: 'aws_s3_acl',
        },
        region: {
                from_secret: 'aws_s3_region',
        },
        bucket: {
                from_secret: 'aws_s3_bucket',
        },
        endpoint: {
                from_secret: 'aws_s3_endpoint',
        },
        path_style: {
                from_secret: 'aws_s3_path_style',
        },
```

**Comment by silverwind** [comment]
We could also just use a JavaScript/Go script to auto-generate a YAML or JSON output file for drone to consume. Arguably better then learning another language.

**Comment by jolheiser** [comment]
> We could also just use a JavaScript/Go script to auto-generate a YAML or JSON output file for drone to consume. Arguably better then learning another language.

Good point. Originally jsonnet seemed good because drone can consume it directly, but during development it seemed better to generate yaml instead as a concrete output. At that point I didn't reconsider that _anything_ could generate the output.

I'll open a new PR for this using Go structs with the yaml encoder.

**Comment by silverwind** [comment]
I'd argue JS would be a better language to generate JSON and in turn YAML than go where JSON data is not so easy to work with (needs `map[string]interface{}` etc unless you want to struct everything).

**Comment by jolheiser** [comment]
Oh I plan on making structs, at least for the things we use (not going to be the entire drone spec)

As for not using JS, I just prefer Go for the types and we already have the yaml module as a dep.

**Comment by wxiaoguang** [comment]
The "drone.yml" has changed a lot, is this PR still related?

**Pull request closed without merging**